File: diffusatory/server/model_profiles.py
from __future__ import annotations

import re
from pathlib import Path
from typing import Iterable, Literal, Protocol

from pydantic import BaseModel
from safetensors import safe_open


ModelFamily = Literal["flux", "sdxl", "unknown"]
ComponentMode = Literal["integrated", "external", "unknown"]
SpeedProfile = Literal["four-step", "standard"]


class CheckpointLike(Protocol):
    title: str
    filename: str


class ModelDefaults(BaseModel):
    steps: int
    sampler: str
    scheduler: str
    cfg_scale: float
    distilled_cfg_scale: float | None = None
    preview_every: int = 5


class ModelProfile(BaseModel):
    checkpoint: str
    family: ModelFamily
    component_mode: ComponentMode
    speed_profile: SpeedProfile
    recommended_modules: list[str]
    defaults: ModelDefaults


FLUX_COMPONENTS = [
    "clip_l.safetensors",
    "t5xxl_fp8_e4m3fn.safetensors",
    "diffusion_pytorch_model.safetensors",
]


def _compact_name(name: str) -> str:
    return re.sub(r"[^a-z0-9]+", "", name.lower())


def _flux_defaults(name: str, four_step: bool) -> ModelDefaults:
    """Return a coherent starting recipe for the installed Flux checkpoint.

    These are model recipes, not universal quality rankings.  Keep the match
    narrow so an unrelated checkpoint never inherits settings just because it
    happens to be Flux.
    """
    compact = _compact_name(name)
    if "c4pacitorxv1" in compact:
        return ModelDefaults(
            steps=42,
            sampler="DEIS",
            scheduler="Beta",
            cfg_scale=1,
            distilled_cfg_scale=3.5,
            preview_every=5,
        )
    if "redcrafthybridh3krea2dualreveal5sfwultra" in compact:
        return ModelDefaults(
            steps=15,
            sampler="DPM++ 2M",
            scheduler="SGM Uniform",
            cfg_scale=1,
            distilled_cfg_scale=3.5,
            preview_every=3,
        )
    if "ultrarealfinetunev4" in compact:
        return ModelDefaults(
            steps=35,
            sampler="DPM++ 2M",
            scheduler="Beta",
            cfg_scale=1,
            distilled_cfg_scale=3,
            preview_every=5,
        )
    return ModelDefaults(
        steps=4 if four_step else 20,
        sampler="Euler",
        scheduler="Simple",
        cfg_scale=1,
        distilled_cfg_scale=3.5,
        preview_every=1 if four_step else 5,
    )


def _checkpoint_keys(filename: str) -> list[str]:
    path = Path(filename)
    if path.suffix.lower() != ".safetensors" or not path.is_file():
        return []
    # ``safe_open`` reads the small tensor index, not the tensor payload. This
    # lets the catalog describe a 12 GB checkpoint without loading 12 GB.
    with safe_open(str(path), framework="pt", device="cpu") as checkpoint:
        return list(checkpoint.keys())


def _family(keys: Iterable[str], name: str) -> ModelFamily:
    keys = tuple(keys)
    if any(
        key.startswith("double_blocks.")
        or key.startswith("model.diffusion_model.double_blocks.")
        for key in keys
    ):
        return "flux"
    if any(key.startswith("conditioner.embedders.") for key in keys) and any(
        key.startswith("model.diffusion_model.") for key in keys
    ):
        return "sdxl"

    lowered = name.lower()
    if "flux" in lowered or "schnell" in lowered:
        return "flux"
    if "sdxl" in lowered or "illustrious" in lowered or "pony" in lowered:
        return "sdxl"
    return "unknown"


def _is_four_step(name: str) -> bool:
    compact = _compact_name(name)
    return (
        "4steps" in compact
        or "foursteps" in compact
        or "schnell" in compact
    )


def profile_checkpoint(checkpoint: CheckpointLike) -> ModelProfile:
    keys = _checkpoint_keys(checkpoint.filename)
    family = _family(keys, checkpoint.title)
    four_step = _is_four_step(checkpoint.title)

    if family == "flux":
        integrated = any(key.startswith("text_encoders.") for key in keys) and any(
            key.startswith("vae.") for key in keys
        )
        component_mode: ComponentMode = "integrated" if integrated else "external"
        modules = [] if integrated else FLUX_COMPONENTS.copy()
        defaults = _flux_defaults(checkpoint.title, four_step)
    elif family == "sdxl":
        component_mode = "integrated"
        modules = []
        defaults = ModelDefaults(
            steps=20,
            sampler="Euler a",
            scheduler="Karras",
            cfg_scale=5,
            preview_every=5,
        )
    else:
        component_mode = "unknown"
        modules = []
        defaults = ModelDefaults(
            steps=20,
            sampler="Euler a",
            scheduler="Karras",
            cfg_scale=5,
            preview_every=5,
        )

    return ModelProfile(
        checkpoint=checkpoint.title,
        family=family,
        component_mode=component_mode,
        speed_profile="four-step" if four_step else "standard",
        recommended_modules=modules,
        defaults=defaults,
    )


def model_profiles(checkpoints: Iterable[CheckpointLike]) -> list[ModelProfile]:
    return [profile_checkpoint(checkpoint) for checkpoint in checkpoints]

File: diffusatory/web/src/domain/modelProfiles.ts
import type {
  ForgeCatalog,
  ModelProfile,
  ModelModule,
} from "../api/forge/types";
import type { GenerationDraft } from "./draft";
import type { SavedModelDefault } from "./savedModelDefaults";

function compact(value: string): string {
  return value.toLowerCase().replace(/[^a-z0-9]+/g, "");
}

function fluxDefaults(name: string, fourStep: boolean): ModelProfile["defaults"] {
  if (name.includes("c4pacitorxv1")) {
    return {
      steps: 42,
      sampler: "DEIS",
      scheduler: "Beta",
      cfg_scale: 1,
      distilled_cfg_scale: 3.5,
      preview_every: 5,
    };
  }
  if (name.includes("redcrafthybridh3krea2dualreveal5sfwultra")) {
    return {
      steps: 15,
      sampler: "DPM++ 2M",
      scheduler: "SGM Uniform",
      cfg_scale: 1,
      distilled_cfg_scale: 3.5,
      preview_every: 3,
    };
  }
  if (name.includes("ultrarealfinetunev4")) {
    return {
      steps: 35,
      sampler: "DPM++ 2M",
      scheduler: "Beta",
      cfg_scale: 1,
      distilled_cfg_scale: 3,
      preview_every: 5,
    };
  }
  return {
    steps: fourStep ? 4 : 20,
    sampler: "Euler",
    scheduler: "Simple",
    cfg_scale: 1,
    distilled_cfg_scale: 3.5,
    preview_every: fourStep ? 1 : 5,
  };
}

function fallbackProfile(checkpoint: string): ModelProfile {
  const name = compact(checkpoint);
  const knownFlux =
    name.includes("c4pacitorxv1") ||
    name.includes("redcrafthybridh3krea2dualreveal5sfwultra") ||
    name.includes("ultrarealfinetunev4");
  const family = name.includes("flux") || name.includes("schnell") || knownFlux
    ? "flux"
    : name.includes("sdxl") || name.includes("illustrious") || name.includes("pony")
      ? "sdxl"
      : "unknown";
  const fourStep = name.includes("4steps") || name.includes("foursteps") || name.includes("schnell");
  const integratedFlux = family === "flux" && (name.includes("aio") || name.includes("allinone"));

  if (family === "flux") {
    return {
      checkpoint,
      family,
      component_mode: integratedFlux ? "integrated" : "external",
      speed_profile: fourStep ? "four-step" : "standard",
      recommended_modules: integratedFlux
        ? []
        : [
            "clip_l.safetensors",
            "t5xxl_fp8_e4m3fn.safetensors",
            "diffusion_pytorch_model.safetensors",
          ],
      defaults: fluxDefaults(name, fourStep),
    };
  }

  return {
    checkpoint,
    family,
    component_mode: family === "sdxl" ? "integrated" : "unknown",
    speed_profile: "standard",
    recommended_modules: [],
    defaults: {
      steps: 20,
      sampler: "Euler a",
      scheduler: "Karras",
      cfg_scale: 5,
      distilled_cfg_scale: null,
      preview_every: 5,
    },
  };
}

export function profileForCheckpoint(
  catalog: ForgeCatalog,
  checkpoint: string,
): ModelProfile {
  return (
    catalog.modelProfiles.find((profile) => profile.checkpoint === checkpoint) ??
    fallbackProfile(checkpoint)
  );
}

function selectNamed<T extends { name: string }>(
  options: T[],
  requested: string,
  fallback: string,
): string {
  const normalized = requested.toLowerCase();
  return (
    options.find((option) => option.name.toLowerCase() === normalized)?.name ??
    options.find((option) => option.name.toLowerCase() === fallback.toLowerCase())?.name ??
    options[0]?.name ??
    requested
  );
}

function selectScheduler(
  catalog: ForgeCatalog,
  requested: string,
  fallback?: string,
): string {
  const normalized = requested.toLowerCase();
  return (
    catalog.schedulers.find(
      (scheduler) =>
        scheduler.name.toLowerCase() === normalized ||
        scheduler.label.toLowerCase() === normalized,
    )?.name ??
    (fallback
      ? catalog.schedulers.find(
          (scheduler) =>
            scheduler.name.toLowerCase() === fallback.toLowerCase() ||
            scheduler.label.toLowerCase() === fallback.toLowerCase(),
        )?.name
      : undefined) ??
    catalog.schedulers[0]?.name ??
    requested
  );
}

function moduleBasename(modelModule: ModelModule): string {
  return modelModule.filename.split(/[\\/]/).at(-1) ?? modelModule.filename;
}

function selectedModuleNames(modules: string[]): Set<string> {
  return new Set(
    modules.flatMap((module) => [module, module.split(/[\\/]/).at(-1) ?? module]),
  );
}

export function recommendedModules(
  catalog: ForgeCatalog,
  profile: ModelProfile,
): string[] {
  return profile.recommended_modules.flatMap((required) => {
    const found = catalog.modules.find(
      (candidate) =>
        candidate.model_name === required || moduleBasename(candidate) === required,
    );
    return found ? [found.filename] : [];
  });
}

export function missingRecommendedModules(
  profile: ModelProfile,
  selectedModules: string[],
): string[] {
  if (profile.component_mode !== "external") return [];
  const selected = selectedModuleNames(selectedModules);
  return profile.recommended_modules.filter((required) => !selected.has(required));
}

export function modelReadinessIssue(
  catalog: ForgeCatalog,
  draft: GenerationDraft,
): string | null {
  const profile = profileForCheckpoint(catalog, draft.checkpoint);
  const missing = missingRecommendedModules(profile, draft.modules);
  if (!missing.length) return null;
  return `This ${profile.family.toUpperCase()} checkpoint still needs ${missing.join(
    ", ",
  )}. Select the required components before generating.`;
}

export function applyCheckpointProfile(
  draft: GenerationDraft,
  catalog: ForgeCatalog,
  checkpoint: string,
  savedDefault?: SavedModelDefault,
): GenerationDraft {
  const profile = profileForCheckpoint(catalog, checkpoint);
  const profiled =
    profile.family === "unknown"
      ? { ...draft, checkpoint }
      : {
          ...draft,
          checkpoint,
          modules: recommendedModules(catalog, profile),
          sampler: selectNamed(
            catalog.samplers,
            profile.defaults.sampler,
            draft.sampler,
          ),
          scheduler: selectScheduler(catalog, profile.defaults.scheduler),
          steps: profile.defaults.steps,
          cfgScale: profile.defaults.cfg_scale,
          distilledCfgScale: profile.defaults.distilled_cfg_scale ?? 3.5,
          previewEvery: profile.defaults.preview_every,
        };
  if (!savedDefault) return profiled;

  const modules = savedDefault.modules.flatMap((savedModule) => {
    const basename = savedModule.split(/[\\/]/).at(-1) ?? savedModule;
    const match = catalog.modules.find(
      (modelModule) =>
        modelModule.filename === savedModule ||
        modelModule.model_name === savedModule ||
        moduleBasename(modelModule) === basename,
    );
    return match ? [match.filename] : [];
  });
  return {
    ...profiled,
    modules,
    sampler: selectNamed(catalog.samplers, savedDefault.sampler, profiled.sampler),
    scheduler: selectScheduler(catalog, savedDefault.scheduler, profiled.scheduler),
    steps: savedDefault.steps,
    cfgScale: savedDefault.cfgScale,
    distilledCfgScale: savedDefault.distilledCfgScale,
    previewEvery: savedDefault.previewEvery,
  };
}
